java/algorithm/src/main/java/org/apache/arrow/algorithm/search/ParallelSearcher.java

```diff
@@ -0,0 +1,122 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.search;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+
+import org.apache.arrow.vector.ValueVector;
+import org.apache.arrow.vector.compare.Range;
+import org.apache.arrow.vector.compare.RangeEqualsVisitor;
+
+/**
+ * Search for a value in the vector by multiple threads.
+ * This is often used in scenarios where the vector is large or
+ * low response time is required.
+ * @param <V> the vector type.
+ */
+public class ParallelSearcher<V extends ValueVector> {
+
+  /**
+   * The target vector to search.
+   */
+  private final V vector;
+
+  /**
+   * The thread pool.
+   */
+  private final ExecutorService threadPool;
+
+  /**
+   * The number of threads to use.
+   */
+  private final int numThreads;
+
+  /**
+   * The position of the key in the target vector, if any.
+   */
+  private int keyPosition = -1;
+
+  /**
+   * Constructs a parallel searcher.
+   * @param vector the vector to search.
+   * @param threadPool the thread pool to use.
+   * @param numThreads the number of threads to use.
+   */
+  public ParallelSearcher(V vector, ExecutorService threadPool, int numThreads) {
+    this.vector = vector;
+    this.threadPool = threadPool;
+    this.numThreads = numThreads;
+  }
+
+  /**
+   * Search for the key in the target vector.
+   * @param keyVector the vector containing the search key.
+   * @param keyIndex the index of the search key in the key vector.
+   * @return the position of a matched value in the target vector,
+   *     or -1 if none is found.
+   * @throws ExecutionException if an exception occurs in a thread.
+   * @throws InterruptedException if a thread is interrupted.
+   */
+  public int search(V keyVector, int keyIndex) throws ExecutionException, InterruptedException {
+    keyPosition = -1;
+    final int valueCount = vector.getValueCount();
+    final CompletableFuture<Boolean>[] futures = new CompletableFuture[numThreads];
+    for (int i = 0; i < futures.length; i++) {
+      futures[i] = new CompletableFuture<>();
+    }
+
+    for (int i = 0; i < numThreads; i++) {
+      final int tid = i;
+      threadPool.submit(() -> {
+        // convert to long to avoid overflow
+        int start = (int) (((long) valueCount) * tid / numThreads);
+        int end = (int) ((long) valueCount) * (tid + 1) / numThreads;
+
+        if (start >= end) {
+          // no data assigned to this task.
+          futures[tid].complete(false);
+          return;
+        }
+
+        RangeEqualsVisitor visitor = new RangeEqualsVisitor(vector, keyVector, false);
+        Range range = new Range(0, 0, 1);
```

Comment:
Why use a range with length one instead of a comparator (like in [`VectorSearcher#linearSearch`](https://github.com/apache/arrow/blob/apache-arrow-0.15.1/java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorSearcher.java#L74))? If one is preferable to the other should we be using it in both places?

Reply:
Good question.
We have two types of semantics for value comparison:
1. equality semantics: the comparison results have two possible values: true (equal), false (un-equal).
2. ordering semantics: the comparison results have 3 possible values: -1 (smaller than), 0 (equal to), and +1 (greater than)

The details can be found in Section 0 of our design [doc](https://docs.google.com/document/d/17nqHWS7gs0vARfeDAcUEbhKMOYHnCtA46TOY_Nls69s/edit?ts=5d825127#heading=h.3432tw3uqfeg).

This particular PR only supports the equality semantics, and support for ordering semantics will be supported in the future. This is because:

1. Small changes are relatively easy to track and review.
2. The support for ordering semantics relies on another feature, which is now in progress (https://github.com/apache/arrow/pull/5659)

Reply:
Understood.